com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/plugin/internal/eval/JdtEvaluationProvider.java

```diff
@@ -68,68 +69,53 @@ public void initialize(IDebugAdapterContext context, Map<String, Object> props)
     }
 
     @Override
-    public CompletableFuture<Value> evaluate(String expression, ThreadReference thread, int depth) {
-        CompletableFuture<Value> completableFuture = new CompletableFuture<>();
-        String projectName = (String) options.get(Constants.PROJECTNAME);
-        if (debugTarget == null) {
-            if (project == null) {
-                if (StringUtils.isBlank(projectName)) {
-                    logger.severe("Cannot evaluate when project is not specified.");
-                    completableFuture.completeExceptionally(new IllegalStateException("Please specify projectName in launch.json."));
-                    return completableFuture;
-                }
-                project = JdtUtils.getJavaProject(projectName);
-            }
-
-            if (project == null) {
-                completableFuture.completeExceptionally(new IllegalStateException(String.format("Project %s cannot be found.", projectName)));
-                return completableFuture;
-            }
-            if (launch == null) {
-                launch = createILaunchMock(project);
-            }
+    public CompletableFuture<Value> evaluateForBreakpoint(IBreakpoint breakpoint, ThreadReference thread, Map<IBreakpoint, Object> breakpointExpressionMap) {
+        if (breakpoint == null) {
+            throw new IllegalArgumentException("breakpoint is null.");
         }
 
-        if (debugTarget == null) {
-            debugTarget = new JDIDebugTarget(launch, thread.virtualMachine(), "", false, false, null, false) {
-                @Override
-                protected synchronized void initialize() {
-                    // use empty initialize intentionally to avoid to register jdi event listener
-                }
-            };
+        if (StringUtils.isBlank(breakpoint.getCondition())) {
+            throw new IllegalArgumentException("breakpoint is not a conditional breakpoint.");
         }
 
-        JDIThread jdiThread = getMockJDIThread(thread);
-        JDIStackFrame stackframe = createStackFrame(jdiThread, depth);
-        if (stackframe == null) {
-            logger.severe("Cannot evaluate because the stackframe is not available.");
-            completableFuture.completeExceptionally(
-                    new IllegalStateException("Cannot evaluate because the stackframe is not available."));
-            return completableFuture;
+        CompletableFuture<Value> failureFuture = new CompletableFuture<>();
+
+        try  {
+            ensureDebugTarget(thread.virtualMachine());
+            JDIThread jdiThread = getMockJDIThread(thread);
+            JDIStackFrame stackframe = (JDIStackFrame) jdiThread.getTopStackFrame();
+
+            ASTEvaluationEngine engine = new ASTEvaluationEngine(project, debugTarget);
+            ICompiledExpression ie = (ICompiledExpression) breakpointExpressionMap
+                    .computeIfAbsent(breakpoint, bp -> engine.getCompiledExpression(bp.getCondition(), stackframe));
+
```

Comment:
Return failureFuture and internalEvaluate result are two different instance. The client cannot hoop up with both future.

Reply:
done with a CompletableFuture parameter in internalEvaluate